core/src/main/java/org/apache/druid/math/expr/Parser.java

```diff
@@ -164,24 +183,141 @@ public static Expr flatten(Expr expr)
   /**
    * Applies a transformation to an {@link Expr} given a list of known (or uknown) multi-value input columns that are
    * used in a scalar manner, walking the {@link Expr} tree and lifting array variables into the {@link LambdaExpr} of
-   * {@link ApplyFunctionExpr} and transforming the arguments of {@link FunctionExpr}
-   * @param expr expression to visit and rewrite
-   * @param bindingsToApply
-   * @return
+   * {@link ApplyFunctionExpr} and transforming the arguments of {@link FunctionExpr} as necessary.
+   *
+   * This function applies a transformation for "map" style uses, such as column selectors, where the supplied
+   * expression will be transformed to return an array of results instead of the scalar result (or appropriately
+   * rewritten into existing apply expressions to produce correct results when referenced from a scalar context).
+   *
+   * This function and {@link #foldUnappliedBindings(Expr, Expr.BindingAnalysis, List, String)} exist to handle
+   * "multi-valued" string dimensions, which exist in a superposition of both single and multi-valued during realtime
+   * ingestion, until they are written to a segment and become locked into either single or multi-valued. This also
+   * means that multi-valued-ness can vary for a column from segment to segment, so this family of transformation
+   * functions exist so that multi-valued strings can be expressed in either and array or scalar context, which is
+   * important because the writer of the query might not actually know if the column is definitively always single or
+   * multi-valued (and it might in fact not be).
+   *
+   * @see #foldUnappliedBindings(Expr, Expr.BindingAnalysis, List, String)
    */
-  public static Expr applyUnappliedBindings(Expr expr, Expr.BindingAnalysis bindingAnalysis, List<String> bindingsToApply)
+  public static Expr applyUnappliedBindings(
```

Comment:
nit: wondering if this series of apply/fold unapplied methods should be somewhere else than `Parser`.

Reply:
Maybe? I originally put it here because this is where `flatten` happens, which is another expression transformation/rewrite that felt the closest to what these methods are doing, but I am open to ideas if there is somewhere else more appropriate.